### ISSUE TITLE
Is there a lib for variable parsing? (feature request?)

### ISSUE THREAD

**Issue by inf3rno** [acuminous/yadda#192]
If we use gherkin with `examples` or just a simple sentence with variables, then we have to define parsers. Is there a lib, we could use to give these parsers instead of defining them manually all the time?
Is there a way to generate a step definition skeleton based on the conversations? It is repetitive too to copy-paste text from feature descriptions to step definitions. I think we are missing these tools.

**Comment by cressie176** [comment]
> If we use gherkin with examples or just a simple sentence with variables, then we have to define parsers. Is there a lib, we could use to give these parsers instead of defining them manually all the time?

Sorry I don't follow, can you provide a code example?


> Is there a way to generate a step definition skeleton based on the conversations? It is repetitive too to copy-paste text from feature descriptions to step definitions. I think we are missing these tools.

This has been [discussed before](https://github.com/acuminous/yadda/issues/89). @simoami created a project called [mimik](https://github.com/simoami/mimik) which can do step generation. I haven't investigated further as I think that a separate tool is the way to go.

**Comment by inf3rno** [comment]
@cressie176 

What I am talking about is https://github.com/acuminous/yadda/blob/master/lib/Dictionary.js#L34

`dictionary.define(term, pattern, converters)`

 - term - the simplest to give it manually
 - pattern - reusable
 - converter - reusable, depends on the pattern

For example:

https://github.com/acuminous/yadda/blob/master/examples/mocha-multi-library/lib/dictionary.js#L8
`new Dictionary().define('NUM', /(\d+)/);`

we could use

```
var ycv = require("yadda-common-variables");
var int = ycv.numeric.postitiveInteger;
new Dictionary().define('NUM', int.pattern, int.toNumber);
```

repeating a lot of regex patterns and converters is error prone.

It would be great if we could give the pattern and the converter in a single config object, so we could use

```
var ycv = require("yadda-common-variables");
new Dictionary().define('NUM', ycv.match.positiveInteger().convert.toNumber());
```

or even

```
new Dictionary().define('NUM').positiveInteger().toNumber();
```

**Comment by cressie176** [comment]
Thanks for explaining the issue, however I don't think it's the right thing to add a predefined dictionary to Yadda. It's simple for project teams (or even companies) to create one and publish it to npm a private repository if they're worried about duplication, and since Yadda supports multiple languages I would have to support many variations of the same term quite ignoring abbreviations and alternatives.

**Comment by inf3rno** [comment]
@cressie176 Ofc. it would be better to put this into a separate library, but what prevents "acuminous" to do it? So it would be an official lib, which could be used by anyone, instead of something what I made in my free time. I'll write something now anyways, because I need it currently, but I am not sure the code quality will be the best. To create more general libs usually you need a lot of time to think about the api. I know, I tried it many times and I failed many times. :-)))

**Comment by cressie176** [comment]
> but what prevents "acuminous" to do it?

Mainly that I don't have any interest in writing it. In the 2-3 years I've been using Yadda and the [various incarnations](http://energizedwork.github.io/selenium-ide-nle/) before it was called that, I haven't used dictionaries for type conversion beyond numbers (which are not always positive), floats (should I use native floats or BigNumber?) and dates (should I use native dates or BigNumber?).

In fact the original use case for dictionaries was to make steps easier to read. I was working on [Visual Hospital](http://lean-health.blogspot.co.uk/2010/02/visual-hospital-system.html) and I needed to differentiate "a $gender patient", (where gender could be male or female) from "a $speciality patient", (where speciality may be one cardiovascular, respiratory, etc). This lead to some ugly regular expressions, which I hid behind dictionaries.

I added the ability to transform the values not because I wanted to parse types, but because I wanted to look up patients by name in later steps

Anyway have fun with [yadda-variables]( https://github.com/inf3rno/yadda-variables).
